### DISCUSSION TITLE
Running behind nginx: UI can't connect via websocket

### DISCUSSION THREAD

**Discussion by garymm** [zwave-js/zwave-js-ui#3245] [Q&A]
Similar to https://github.com/zwave-js/zwave-js-ui/discussions/1396, but not sure if I should post there or start a new one.

I've got zwave-js-ui running in a docker container with ports 8091 and 3000 exposed.
I've got nginx configured and successfully loading the UI, but the UI can't connect via websocket.
I see these errors in the chrome console:
```
WebSocket connection to 'wss://zwavejs.example.com/socket.io/?token=...' failed: WebSocket is closed before the connection is established.
```

Here's my nginx configuration:

```
upstream zwavejs {
  server <redacted>:8091;
}

map $http_upgrade $connection_upgrade {
    default upgrade;
    '' close;
}

server {
  listen              443 ssl;
  server_name         zwavejs.example.com;

  location / {
      proxy_pass         https://zwavejs;
      proxy_read_timeout 120s;
      proxy_redirect off;
      proxy_buffering off;
      proxy_set_header        Host              $host;
      proxy_set_header        X-Real-IP         $remote_addr;
      proxy_set_header        X-Forwarded-For   $proxy_add_x_forwarded_for;
      proxy_set_header        X-Forwarded-Proto $scheme;
      # https://zwave-js.github.io/zwave-js-ui/#/usage/reverse-proxy?id=accessing-from-behind-a-proxy
      proxy_set_header        Upgrade           $http_upgrade;
      proxy_set_header        Connection        $connection_upgrade;
  }
}
```

I have https enabled in the UI settings, and I've set `SSL_CERTIFICATE` and `SSL_KEY` environment variables.
The app prints to stdout that it's using HTTPS and has loaded the cert and key.

I've also tried changing the proxy_pass to `http://zwavejs` and disabling HTTPS (but keeping HTTPS on nginx), but that didn't work either.

Things work fine when I access it via http://localhost:8091.

Any tips?

**Comment by kpine** [comment]
One possibility is because you are missing the `proxy_http_version 1.1;` setting as part of the websocket configuration. It's included in both the [ZUI example](https://zwave-js.github.io/zwave-js-ui/#/usage/reverse-proxy?id=accessing-from-behind-a-proxy) you have listed, and also [Nginx WebSocket proxying](https://nginx.org/en/docs/http/websocket.html) docs. According to MDN, the [Upgrade header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Upgrade) is only valid for HTTP 1.1, and the [nginx default](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_http_version) is 1.0.

**Reply by garymm** [reply]
Thanks for taking a look! I tried, but that didn't seem to fix anything.

**Reply by kpine** [reply]
What if you disable SSL in ZUI now?

**Comment by robertsLando** [comment]
Here you can find a working example: https://github.com/zwave-js/zwave-js-ui/tree/master/docker/nginx
